src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,436 @@
+            // Mt = wrap_angle(Mt);
+            return solve_kepler_hyperbolic(Mt, m_eccentricity, 20000);
+#else // Use universal kepler solver
+            double chi = solve_kepler_universal(m_r0, m_v0, m_semiMajorAxis, time - m_epoch, m_gravitationalParameter, 2000);
+            double F = chi / std::sqrt(-m_semiMajorAxis) + m_F0;
+            return F;
+#endif
+        }
+    }
+
+    void KeplerOrbit::get_fixed_reference_frame(Vector3d &radial, Vector3d &transverse, Vector3d &outOfPlane) const
+    {
+        double a = m_argumentOfPeriapsis;
+        double b = m_longitudeOfAscendingNode;
+        double c = m_inclination;
+
+        radial = Vector3d(std::cos(a) * std::cos(b) - std::sin(a) * std::sin(b) * std::cos(c),
+                          std::cos(a) * std::sin(b) + std::sin(a) * std::cos(b) * std::cos(c),
+                          std::sin(a) * std::sin(c));
+
+        transverse = Vector3d(-std::sin(a) * std::cos(b) - std::cos(a) * std::sin(b) * std::cos(c),
+                              -std::sin(a) * std::sin(b) + std::cos(a) * std::cos(b) * std::cos(c),
+                              std::cos(a) * std::sin(c));
+
+        outOfPlane = Vector3d(std::sin(b) * std::sin(c),
+                              -std::cos(b) * std::sin(c),
+                              std::cos(c));
+    }
+
+    void KeplerOrbit::get_state_vectors_at_time(double time, Vector3d &position, Vector3d &velocity) const
+    {
+        double E = get_eccentric_anomaly(time);
+        double trueAnomaly = get_true_anomaly_from_eccentric(E);
+        get_state_vectors_at_eccentric_anomaly(E, position, velocity);
+    }
+
+    void KeplerOrbit::get_state_vectors_at_eccentric_anomaly(double E, Vector3d &position, Vector3d &velocity) const
+    {
+        double trueAnomaly = get_true_anomaly_from_eccentric(E);
+        double r = get_orbiting_radius(m_eccentricity, m_semiMajorAxis, trueAnomaly);
+        Vector3d radial, transverse, outOfPlane;
+        get_fixed_reference_frame(radial, transverse, outOfPlane);
+
+        position = r * (std::cos(trueAnomaly) * radial + std::sin(trueAnomaly) * transverse);
+        if (m_eccentricity < 1.0)
+        {
+            velocity = (std::sqrt(m_gravitationalParameter * m_semiMajorAxis) / r) * (-std::sin(E) * radial + (std::sqrt(1 - m_eccentricity * m_eccentricity) * std::cos(E)) * transverse);
+        }
+        else
+        {
+            double v_r = (m_gravitationalParameter / m_h * m_eccentricity * std::sin(trueAnomaly));
+            double v_perp = (m_h / r);
+            Vector3d radusDir = position.normalized();
+            Vector3d perpDir = cross(outOfPlane, radusDir).normalized();
+
+            velocity = v_r * radusDir + v_perp * perpDir;
+        }
+    }
+
+    void KeplerOrbit::rebase_epoch(double newEpoch)
+    {
+        double new_E0 = get_eccentric_anomaly(newEpoch);
+        double new_m0 = new_E0 - m_eccentricity * std::sin(new_E0);
+        Vector3d newPos, newVel;
```

Comment:
i strongly discourage declaring more than one variable per line, with a comma like this.
It's very prone to confusing junior programmers.

Only exception would be situations like in a for-loop where the comma is the only way to keep the scoping concise and still get multiple variables.